docs/proposals/202003_thanos_rules_federation.md

```diff
@@ -0,0 +1,218 @@
+Note in the example above how the recording rule `r1` is not deduplicated as it is contained in two different groups.
+
+Scenario 2:
+
+The next level of deduplication is governed by the label/value set of the underlying recording/alerting rule while respecting the replica label.
+For a given conflict, the youngest is preferred. For alerting rules, the youngest firing rule is preferred.
+
+Given the following stream of incoming recording rules:
+```text
+group: a
+   recording:<name:"r1" labels:<labels:<name:"replica" value:"thanos-ruler-1" > > last_evaluation:<seconds:1 > >
+group: a
+   recording:<name:"r1" labels:<labels:<name:"replica" value:"thanos-ruler-2" > > last_evaluation:<seconds:2 > >
+```
+
+The output becomes:
+```text
+group: a
+   recording:<name:"r1" labels:<labels:<name:"replica" value:"thanos-ruler-2" > > last_evaluation:<seconds:2 > >
+```
+
+Given the following stream of incoming alerting rules:
+```text
+group: a
+   alert:<state:FIRING name:"a1" labels:<labels:<name:"replica" value:"thanos-ruler-1" > > last_evaluation:<seconds:1 > >
+group: a
+   alert:<             name:"a1" labels:<labels:<name:"replica" value:"thanos-ruler-2" > > last_evaluation:<seconds:2 > >
+```
+
+The output becomes:
+```text
+group: a
+   alert:<state:FIRING name:"a1" labels:<labels:<name:"replica" value:"thanos-ruler-1" > > last_evaluation:<seconds:1 > >
+```
+
+Note how in the above output the firing alerting rule was preferred despite being older.
+
+Scenario 3:
+
+If, under the above conditions a rule is a candidate for deduplication, finally the rule `expr` and `for` fields are being considered for deduplication.
+
+Given the following stream of incoming alerting rules will also result in two independent alerting rules as both the `expr` and `for` fields differ:
+```text
+  - alert: KubeAPIErrorBudgetBurn
+    annotations:
+      message: The API server is burning too much error budget
+    expr: |
+      sum(apiserver_request:burnrate1h) > (14.40 * 0.01000)
+      and
+      sum(apiserver_request:burnrate5m) > (14.40 * 0.01000)
+    for: 2m
+    labels:
+      severity: critical
+  - alert: KubeAPIErrorBudgetBurn
+    annotations:
+      message: The API server is burning too much error budget
+    expr: |
+      sum(apiserver_request:burnrate6h) > (6.00 * 0.01000)
+      and
+      sum(apiserver_request:burnrate30m) > (6.00 * 0.01000)
+    for: 15m
+    labels:
+      severity: critical
+```
+
+## Alternatives
+
+* Cortex contains a sharded Ruler. Assigning rules to shards is done via Consul, though a gossip implementation is under development. Shards do not communicate with other shards. Rules come from a store (e.g. a Postgres database).
+
+## Work Plan
+
+* Implement a new flag `--rule` in Thanos Querier which registers RulesAPI endpoints.
+* Implement a new flag `--rule.replica-label` in Thanos Querier.
+* Implement RulesAPI backends in sidecar, query, rule.
+* Feature branch: https://github.com/thanos-io/thanos/pull/2200
+
+## Future
+
```

Comment:
Let's add sentence to clarify this:

These changes are suggestions which we will need to be discussed in future and are not part of the proposal implementation.

Reply:
fixed :+1: